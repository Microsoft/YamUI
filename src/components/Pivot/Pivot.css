/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';
@import 'src/css/variables/fonts.css';

.y-pivot .ms-Pivot-link {
  @extend text__small;
  color: $color__deathMetal;
  font-weight: $fontWeight__bold;
  height: 28px;
  opacity: 0.7;
  padding-bottom: 0.8rem;
}

.y-pivot .ms-Pivot-link.is-selected {
  opacity: 1;
}

.y-pivot .ms-Pivot-link.is-selected::before {
  border-bottom-color: $color__river;
}

.y-pivot.y-pivot__isDark .ms-Pivot-link {
  color: $color__white;
}

.y-pivot.y-pivot__isDark .ms-Pivot-link.is-selected::before {
  border-bottom-color: $color__white;
}
