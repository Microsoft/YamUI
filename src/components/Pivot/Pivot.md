### Examples

Basic usage:

```js { "props": { "data-description": "basic" } }
const pivotItems = [{ text: 'DISCOVERY', key: '0' }, { text: 'ALL', key: '1' }, { text: 'FOLLOWING', key: '2' }];

<Pivot onChange={action('onChange')} pivotItems={pivotItems} />;
```

With Dark theme:

```js { "props": { "data-description": "with dark background" } }
const pivotItems = [
  { text: 'NEW CONVERSATIONS', key: '0' },
  { text: 'ALL CONVERSATIONS', key: '1' },
  { text: 'FILES', key: '2' },
];
<div style={{ backgroundColor: '#386cbb' }}>
  <Pivot onChange={action('onChange')} overDarkBackground={true} pivotItems={pivotItems} />
</div>;
```

With itemCount prop:

```js { "props": { "data-description": "with itemCount prop in pivotItems" } }
const pivotItems = [
  { text: 'USERS', key: '0', itemCount: 1 },
  { text: 'GROUPS', key: '1', itemCount: 0 },
  { text: 'FILES', key: '2', itemCount: 5437253445 },
];

<Pivot onChange={action('onChange')} pivotItems={pivotItems} />;
```
