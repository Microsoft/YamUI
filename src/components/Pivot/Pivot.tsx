/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { Pivot as FabricPivot, PivotItem as FabricPivotItem } from 'office-ui-fabric-react/lib/Pivot';

import './Pivot.css';

export interface PivotItem {
  ariaLabel?: string;
  key: string;
  text: string;
}

export interface PivotProps extends NestableBaseComponentProps {
  /**
   * Indicates whether the text on the tabs will be rendered on white colors
   * to be used with dark backgrounds.
   */
  isDark?: boolean;
  /**
   * Indicates the key of the initially selected tab or pivot item.
   */
  selectedKey: string;
  /**
   * Contains the information of the pivot items that will be rendered within this component.
   */
  pivotItems: PivotItem[];
  /**
   * Callback that will be executed every time there is a change in the selected pivot item.
   */
  onChange(key: string): void;
}

export default class Pivot extends React.Component<PivotProps> {
  public render() {
    return (
      <FabricPivot
        className={this.getClasses()}
        selectedKey={this.props.selectedKey}
        onLinkClick={this.onLinkClick}
        headersOnly={true}
      >
        {this.getFabricPivotItems()}
      </FabricPivot>
    );
  }

  private getFabricPivotItems() {
    return this.props.pivotItems.map((child: PivotItem) => (
      <FabricPivotItem ariaLabel={child.ariaLabel} itemKey={child.key} headerText={child.text} key={child.key} />
    ));
  }

  private onLinkClick = (item?: FabricPivotItem) => {
    if (item && item.props.itemKey && item.props.itemKey !== this.props.selectedKey) {
      this.props.onChange(item.props.itemKey);
    }
  };

  private getClasses() {
    const classes: string[] = ['y-pivot'];
    const { className, isDark } = this.props;

    if (isDark) {
      classes.push('y-pivot__isDark');
    }
    if (className) {
      classes.push(className);
    }

    return classes.join(' ');
  }
}
