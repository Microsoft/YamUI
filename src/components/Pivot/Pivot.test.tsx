/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Pivot, { PivotProps, PivotItem } from './index';

describe('<Pivot />', () => {
  let component: ShallowWrapper<PivotProps>;
  let onChange: jest.Mock<Function>;

  const getComponent = (overrides: Partial<PivotProps>) => {
    const pivotItems: PivotItem[] = [
      { key: '1', text: 'Tab 1', ariaLabel: 'test' },
      { key: '2', text: 'Tab 2' },
      { key: '3', text: 'Tab 3' },
    ];

    return shallow(<Pivot selectedKey="1" pivotItems={pivotItems} onChange={jest.fn()} {...overrides} />);
  };

  describe('with default options', () => {
    beforeEach(() => {
      onChange = jest.fn();
      component = getComponent({ onChange });
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('when an already selected tab is clicked', () => {
      beforeEach(() => {
        component.find('StyledPivotBase').simulate('linkClick', {
          props: { itemKey: '1' },
        });
      });

      it('calls onChange prop', () => {
        expect(onChange).not.toHaveBeenCalled();
      });
    });

    describe('when a non selected tab is clicked', () => {
      beforeEach(() => {
        component.find('StyledPivotBase').simulate('linkClick', {
          props: { itemKey: '2' },
        });
      });

      it('calls onChange prop', () => {
        expect(onChange).toHaveBeenCalled();
      });
    });
  });

  describe('with isDark as true', () => {
    beforeEach(() => {
      component = getComponent({ isDark: true });
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
