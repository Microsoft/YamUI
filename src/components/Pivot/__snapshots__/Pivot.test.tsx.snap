// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pivot /> with default options matches its snapshot 1`] = `
<StyledPivotBase
  className="y-pivot"
  headersOnly={true}
  onLinkClick={[Function]}
  selectedKey="1"
>
  <PivotItem
    ariaLabel="test"
    headerText="Tab 1"
    itemKey="1"
    key="1"
  />
  <PivotItem
    headerText="Tab 2"
    itemKey="2"
    key="2"
  />
  <PivotItem
    headerText="Tab 3"
    itemKey="3"
    key="3"
  />
</StyledPivotBase>
`;

exports[`<Pivot /> with isDark as true matches its snapshot 1`] = `
<StyledPivotBase
  className="y-pivot y-pivot__isDark"
  headersOnly={true}
  onLinkClick={[Function]}
  selectedKey="1"
>
  <PivotItem
    ariaLabel="test"
    headerText="Tab 1"
    itemKey="1"
    key="1"
  />
  <PivotItem
    headerText="Tab 2"
    itemKey="2"
    key="2"
  />
  <PivotItem
    headerText="Tab 3"
    itemKey="3"
    key="3"
  />
</StyledPivotBase>
`;
